### ISSUE TITLE
Exclude _anonymous from external authentication

### ISSUE THREAD

**Issue by Shaoranlaos** [scm-manager/scm-manager#1548]
#### Problem to be solved

We have enabled anonymous access via protocol but not via Web-Frontend and always have some (3-4) seconds delay before the login page is shown.
The follwing log line indicates the root of the problem:
WARN  sonia.scm.auth.ldap.UserSearcher - no user with username _anonymous found

Imho the internal _anonymous user should not be authenticated against an external provider, in our case an AD-Server.

#### Proposed solution

Exclude the _anonymous user from the external authentication resolving. Or if that is to extreme provide a config option to select if it should try to resolve it.
That would reduce the waiting time for the end user by some seconds (in our case 3-4s) before he/she/it can log into the GUI.

**Comment by sdorra** [comment]
The LdapRealm should not try to authenticate the `_anonymous` User, because the anonymous authentication uses a different type of token as the ldap authentication. I don't think that the problem is the authentication, but i think the resolving of the groups could be the problem. This is a known problem and it is already fixed for the upcoming version of SCM-Manager. For more information have a look at #1541

Just to be sure that it is the group resolver problem, could you please do the following steps:

* install the support plugin
* start log recording
* authenticate with an ad user
* stop log recording
* post the created zip

**Comment by Shaoranlaos** [comment]
I'm a bit to uncomfortable with posting the whole log/zip here because of the information in it (and i am not confident i can redact all of it) but i think the following is the relevant part of the authentication process of the _anonymous user

```
...
08:57:24.496 [qtp1305193908-36] TRACE sonia.scm.api.v2.CacheControlResponseFilter - add no-cache header to response
08:57:24.526 [qtp1305193908-35] TRACE sonia.scm.web.filter.AuthenticationFilter - anonymous access granted
08:57:24.527 [qtp1305193908-35] DEBUG sonia.scm.security.DAORealmHelper - try to authenticate _anonymous
08:57:24.527 [qtp1305193908-35] TRACE sonia.scm.security.DefaultKeyGenerator - create new key ADSPRqj4Jp
08:57:24.528 [qtp1305193908-35] TRACE sonia.scm.event.LegmanScmEventBus - register sonia.scm.api.v2.resources.ScmPathInfoStore@523e48a4 to event bus ScmEventBus-1
08:57:24.528 [qtp1305193908-35] TRACE sonia.scm.event.LegmanScmEventBus - register sonia.scm.api.v2.resources.ResourceLinks@68e80078 to event bus ScmEventBus-1
08:57:24.528 [qtp1305193908-35] TRACE sonia.scm.event.LegmanScmEventBus - register sonia.scm.api.v2.resources.IndexDtoGenerator@35a79c09 to event bus ScmEventBus-1
08:57:24.528 [qtp1305193908-35] TRACE sonia.scm.event.LegmanScmEventBus - register sonia.scm.api.v2.resources.IndexResource@6a6a6859 to event bus ScmEventBus-1
08:57:24.528 [qtp1305193908-35] DEBUG sonia.scm.security.SecurityRequestFilter - allowed unauthenticated request to method public sonia.scm.api.v2.resources.IndexDto sonia.scm.api.v2.resources.IndexResource.getIndex()
08:57:24.529 [qtp1305193908-35] DEBUG sonia.scm.auth.ldap.LdapConnection - create context for dn ****
08:57:24.529 [qtp1305193908-35] DEBUG sonia.scm.auth.ldap.LdapConnection - use follow as referral strategy
08:57:24.533 [qtp1305193908-35] DEBUG sonia.scm.auth.ldap.UserSearcher - using scope sub for user search
08:57:24.533 [qtp1305193908-35] DEBUG sonia.scm.auth.ldap.UserSearcher - search-filter for user search: (&(objectClass=Person)(sAMAccountName=_anonymous))
08:57:24.590 [CronScheduler-1-2] TRACE sonia.scm.schedule.CronTask - skip execution of task sonia.scm.repository.GitGcTask, because expression 0 0 2 ? * SUN does not match
08:57:25.222 [CronScheduler-1-1] TRACE sonia.scm.schedule.CronTask - skip execution of task com.cloudogu.scm.review.events.ChannelCleanupTask, because expression 0/30 * * * * ? does not match
08:57:25.590 [CronScheduler-1-2] TRACE sonia.scm.schedule.CronTask - skip execution of task sonia.scm.repository.GitGcTask, because expression 0 0 2 ? * SUN does not match
08:57:26.222 [CronScheduler-1-1] TRACE sonia.scm.schedule.CronTask - skip execution of task com.cloudogu.scm.review.events.ChannelCleanupTask, because expression 0/30 * * * * ? does not match
08:57:26.590 [CronScheduler-1-2] TRACE sonia.scm.schedule.CronTask - skip execution of task sonia.scm.repository.GitGcTask, because expression 0 0 2 ? * SUN does not match
08:57:27.222 [CronScheduler-1-1] TRACE sonia.scm.schedule.CronTask - skip execution of task com.cloudogu.scm.review.events.ChannelCleanupTask, because expression 0/30 * * * * ? does not match
08:57:27.590 [CronScheduler-1-2] TRACE sonia.scm.schedule.CronTask - skip execution of task sonia.scm.repository.GitGcTask, because expression 0 0 2 ? * SUN does not match
08:57:28.222 [CronScheduler-1-1] TRACE sonia.scm.schedule.CronTask - skip execution of task com.cloudogu.scm.review.events.ChannelCleanupTask, because expression 0/30 * * * * ? does not match
08:57:28.590 [CronScheduler-1-2] TRACE sonia.scm.schedule.CronTask - skip execution of task sonia.scm.repository.GitGcTask, because expression 0 0 2 ? * SUN does not match
08:57:29.222 [CronScheduler-1-1] TRACE sonia.scm.schedule.CronTask - skip execution of task com.cloudogu.scm.review.events.ChannelCleanupTask, because expression 0/30 * * * * ? does not match
08:57:29.590 [CronScheduler-1-2] TRACE sonia.scm.schedule.CronTask - skip execution of task sonia.scm.repository.GitGcTask, because expression 0 0 2 ? * SUN does not match
08:57:30.177 [qtp1305193908-35] WARN  sonia.scm.auth.ldap.UserSearcher - no user with username _anonymous found
08:57:30.177 [qtp1305193908-35] DEBUG sonia.scm.group.DefaultGroupCollector - collected following groups for principal _anonymous: []
08:57:30.177 [qtp1305193908-35] TRACE sonia.scm.security.DefaultAuthorizationCollector - collect AuthorizationInfo for user _anonymous
08:57:30.177 [qtp1305193908-35] TRACE sonia.scm.store.JAXBConfigurationEntryStore - get all items from configuration store
08:57:30.177 [qtp1305193908-35] TRACE sonia.scm.store.FileBasedStore - try to retrieve item with id custom
...```

To me it seems it realy tries to authenticate the _anonymous user against ldap (due to the building of the search filter for ldap) and it takes about 6 seconds.

If you do need the whole zip could you provide me an alternative means to get it to you?

**Comment by sdorra** [comment]
Not necessarily the UserSearcher is also used to find the dn of the user to resolve the groups. I'm not 100% sure if this is the case. It is possible that you install a snapshot version of SCM-Manager and the ldap plugin on a separate machine to test the behaviour with the upcoming version? How did you installed SCM-Manager (docker, windows, rpm, deb)?

**Comment by Shaoranlaos** [comment]
I installed it via rpm on a CentOS 7.
Do i have to compile it myself or is there a link where i can download the snapshot version?
I can try it out on Monday.

**Comment by sdorra** [comment]
You can download the latest snapshots from our [snapshot repository](https://packages.scm-manager.org/service/rest/repository/browse/yum-v2-snapshots/) e.g.: [scm-server-2.13.1.202102101048802.noarch.rpm](https://packages.scm-manager.org/repository/yum-v2-snapshots/scm-server-2.13.1.202102101048802.noarch.rpm).

**Comment by Shaoranlaos** [comment]
OK I tried your linked version and see the same log lines in the log as with 2.13.0

```
07:20:37.956 [qtp237061348-14] DEBUG sonia.scm.security.DAORealmHelper - try to authenticate _anonymous
07:20:37.956 [qtp237061348-14] TRACE sonia.scm.security.DefaultKeyGenerator - create new key 2WSPizwH61Z
07:20:37.957 [qtp237061348-14] TRACE sonia.scm.event.LegmanScmEventBus - register sonia.scm.api.v2.resources.ScmPathInfoStore@68b811a1 to event bus ScmEventBus-1
07:20:37.957 [qtp237061348-14] TRACE sonia.scm.event.LegmanScmEventBus - register sonia.scm.api.v2.resources.ResourceLinks@3d19c5c0 to event bus ScmEventBus-1
07:20:37.957 [qtp237061348-14] TRACE sonia.scm.event.LegmanScmEventBus - register sonia.scm.api.v2.resources.IndexDtoGenerator@617ea319 to event bus ScmEventBus-1
07:20:37.957 [qtp237061348-14] TRACE sonia.scm.event.LegmanScmEventBus - register sonia.scm.api.v2.resources.IndexResource@1567f298 to event bus ScmEventBus-1
07:20:37.957 [qtp237061348-14] DEBUG sonia.scm.security.SecurityRequestFilter - allowed unauthenticated request to method public sonia.scm.api.v2.resources.IndexDto sonia.scm.api.v2.resources.IndexResource.getIndex()
07:20:37.958 [qtp237061348-14] DEBUG sonia.scm.auth.ldap.LdapConnection - create context for dn CN=DRS-AD-Reader,OU=Service-User,OU=Users,OU=DRS,DC=xfab,DC=ads
07:20:37.958 [qtp237061348-14] DEBUG sonia.scm.auth.ldap.LdapConnection - use follow as referral strategy
07:20:37.960 [qtp237061348-16] TRACE sonia.scm.security.DefaultKeyGenerator - create new key BFSPizwHA1a
07:20:37.960 [qtp237061348-16] TRACE sonia.scm.WebResourceServlet - try to load /locales/en/commons.json
07:20:37.960 [qtp237061348-16] TRACE sonia.scm.plugin.DefaultUberWebResourceLoader - return web resource /locales/en/commons.json from cache
07:20:37.960 [qtp237061348-16] TRACE sonia.scm.WebResourceServlet - found /locales/en/commons.json -- serve as resource file:/var/cache/scm/work/webapp/webapp/locales/en/commons.json
07:20:37.960 [qtp237061348-16] DEBUG sonia.scm.WebResourceServlet - using sender to serve /scm/locales/en/commons.json
07:20:37.960 [qtp237061348-14] DEBUG sonia.scm.auth.ldap.UserSearcher - using scope sub for user search
07:20:37.960 [qtp237061348-14] DEBUG sonia.scm.auth.ldap.UserSearcher - search-filter for user search: (&(objectClass=Person)(sAMAccountName=_anonymous))
07:20:37.971 [qtp237061348-17] TRACE sonia.scm.security.DefaultKeyGenerator - create new key 3kSPizwHL1b
07:20:37.972 [qtp237061348-17] TRACE sonia.scm.WebResourceServlet - try to load /assets/tokenizer.bundle.worker.js
07:20:37.972 [qtp237061348-17] TRACE sonia.scm.plugin.DefaultUberWebResourceLoader - return web resource /assets/tokenizer.bundle.worker.js from cache
07:20:37.972 [qtp237061348-17] TRACE sonia.scm.WebResourceServlet - found /assets/tokenizer.bundle.worker.js -- serve as resource file:/var/cache/scm/work/webapp/webapp/assets/tokenizer.bundle.worker.js
07:20:37.972 [qtp237061348-17] DEBUG sonia.scm.WebResourceServlet - using sender to serve /scm/assets/tokenizer.bundle.worker.js
07:20:37.978 [qtp237061348-16] TRACE sonia.scm.security.DefaultKeyGenerator - create new key CBSPizwHS1c
07:20:37.978 [qtp237061348-16] TRACE sonia.scm.WebResourceServlet - try to load /locales/de/commons.json
07:20:37.978 [qtp237061348-16] TRACE sonia.scm.plugin.DefaultUberWebResourceLoader - return web resource /locales/de/commons.json from cache
07:20:37.978 [qtp237061348-16] TRACE sonia.scm.WebResourceServlet - found /locales/de/commons.json -- serve as resource file:/var/cache/scm/work/webapp/webapp/locales/de/commons.json
07:20:37.978 [qtp237061348-16] DEBUG sonia.scm.WebResourceServlet - using sender to serve /scm/locales/de/commons.json
07:20:37.984 [qtp237061348-17] TRACE sonia.scm.security.DefaultKeyGenerator - create new key FYSPizwHY1d
07:20:37.984 [qtp237061348-17] TRACE sonia.scm.WebResourceServlet - try to load /favicon.ico
07:20:37.984 [qtp237061348-17] TRACE sonia.scm.plugin.DefaultUberWebResourceLoader - return web resource /favicon.ico from cache
07:20:37.984 [qtp237061348-17] TRACE sonia.scm.WebResourceServlet - found /favicon.ico -- serve as resource file:/var/cache/scm/work/webapp/webapp/favicon.ico
07:20:37.984 [qtp237061348-17] DEBUG sonia.scm.WebResourceServlet - using sender to serve /scm/favicon.ico
07:20:47.969 [qtp237061348-14] WARN  sonia.scm.auth.ldap.UserSearcher - no user with username _anonymous found
07:20:47.970 [qtp237061348-14] DEBUG sonia.scm.group.DefaultGroupCollector - collected following groups for principal _anonymous: []
07:20:47.970 [qtp237061348-14] TRACE sonia.scm.security.DefaultAuthorizationCollector - collect AuthorizationInfo for user _anonymous
07:20:47.970 [qtp237061348-14] TRACE sonia.scm.store.JAXBConfigurationEntryStore - get all items from configuration store
07:20:47.970 [qtp237061348-14] TRACE sonia.scm.security.DefaultRealm - principal does not contain scope information, returning all permissions
07:20:47.970 [qtp237061348-14] TRACE sonia.scm.security.DefaultRealm - authorization summary: 
username   : _anonymous
roles      : 
 - user
scope      : 
permissions: 
 - user:read:_anonymous
```

So it is again trying to resolve _anonymous in ldap.

I than tried the newest SNAPSHOT (used scm-server-2.13.1.202102190913104.noarch.rpm to upgrade from the previous one) and couldn't find that LDAP search string in the log but that might be due to caching. Is the cache emptied on restart?
I see `07:37:10.493 [qtp237061348-36] TRACE sonia.scm.security.DefaultAuthorizationCollector - retrieve AuthorizationInfo for user _anonymous from cache` in the log.

**Comment by sdorra** [comment]
That is really strange. I've done some debugging this morning. I've first start with the old version 2.13.0 in this version the _anonymous user is searched by the group resolver, never for authentication only for the groups. After that i've used the the version from the develop branch and there was not a single request against the ldap for the _anonymous user.

> Is the cache emptied on restart?
Yes the cache is in memory, so if you restart the service everything is cleared.

> I think this cache access was the whole point of this change.
No, the point of the change is that external group resolvers such as ldap are not called for system accounts such as _anonymous.

**Comment by Shaoranlaos** [comment]
When was the change made? When i used the latest snapshot in the second attempt (the first attempt was with your linked snapshot from the 10th of February) there was no descernable query against ldap even after a restart so i think it is working.

Sorry if I have written my previous comment a bit confusing with my 2 attempts.

With this i think the issue can be closed and I will wait for 2.13.1.
